### PR TITLE
Rework error handling and state flow in the Azure connection

### LINKED ISSUE #4803: Azure Shell can't connect to the internet
# Environment

```none
Microsoft Windows [Version 10.0.18363.657]
Windows Terminal version: 0.9.433.0

Any other software? No
```

# Steps to reproduce
Install from store
Open the Azure Cloud Shell

# Expected behavior
Can connect to Azure Cloud Shell

# Actual behavior

Receive the message: `Could not connect to Azure. You may not have internet or the server might be down.`

### LINKED ISSUE #5325: No way to Rename / Name Tenants
Using Azure Cloudshell with Terminal it brilliantly can save tenants (thank you) however It does the naming on its own, and to the best of my knowledge we can not rename them later, working as a Devops engineer for a large MSP, makes this less than ideal scenario. Esp across devices where the first usage and thus named Default Directory is likely not even the same.  Additionally if there are mutiple accounts with in an organization (lets say a company has an Azure AD domain, and multiple employees have an azure account, as well as a main organization azure account, well I'm sure you get the idea)  Perhaps we could add this to the existing profile.json where names could be edited ?

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
### 38429b56c - az: rework tenant/tenant storage and fix display names
Instead of passing around four loose variables, create a Tenant class
and use that for packing/unpacking into/out of json (and the windows
credential store, where we "cleverly" used json for the tenant info
there too).

When displaying a tenant, use its display name if there is one, the
unknown resource string if there isn't, and the default domain if there
is one and the ID if there isn't.

Fixes #5325.

### b5e16c27c - az: show more tenant info
Switch to 2020 tenant API; use defaultDomain; put defaultDomain into the
display name.

Consider: not showing the tenant GUID at all?

### 61f06e7aa - az: rework error handling
* _RequestHelper no longer eats exceptions.
* Delete the "no internet" error message.
* Wrap exceptions coming out of Azure API in a well-known type.
* Catch by type.
* Extract error codes for known failures (keep polling, invalid grant).
* When we get an Invalid Grant, dispose of the cached refresh token and force the user to log in again.
* Catch all printable exceptions and print them.
* Remove the NoConnect state completely -- just bail out when an exception hits the toplevel of the output thread.
* Move 3x logic into _RefreshTokens and pop exceptions out of it.
* Begin abstracting into AzureClient

## Summary of the Pull Request

This is a tidying-up of the azure connection. It improves error logging (like: it actually emits error logs...) and retry logic and the state machine and it audits the exit points of the state machine for exceptions and removes the HRESULT returns (so they either succeed and transition to a new state or throw an exception or are going down anyway).

There's also a change in here that changes how we display tenants. It adds the "default domain" to the name, so that instead of seeing this:

```
Conhost (aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa)
Default Directory (bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb)
```

you see this

```
Conhost (conhost.onmicrosoft.com)
Default Directory (dustinhowett.onmicrosoft.com)
```

# References

Fixes #5325 (by addressing its chief complaint).
Fixes #4803.
Adds diagnosability for #4575.

## PR Checklist
* [x] Closes #5325. Closes #4803.
* [x] CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] Core contributor

## Detailed Description of the Pull Request / Additional comments
As above.

## Validation Steps Performed
Logged in, out, around a good number of times.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
nts: we can remove AzureExitStr as well.

**Comment by DHowett-MSFT** [comment]
Confirmed this actually _fixes_ #4803, doesn’t just improve diagnostics.

File: src/cascadia/TerminalConnection/AzureConnection.cpp [hunk 1]
```diff
@@ -91,19 +83,48 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
 
+    // Method description:
+    // - helper that prints exception information to the output stream.
+    // Arguments:
+    // - [IMPLICIT] the current exception context
+    void AzureConnection::_WriteCaughtExceptionRecord()
+    {
+        try
+        {
+            throw;
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/AzureConnection.cpp:94]: what the heck

does this just get the current exception if you're already in a `catch` block?

**Comment by DHowett-MSFT** [comment]
Yeah def

File: src/cascadia/TerminalConnection/AzureConnection.cpp (hunk 1 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/AzureConnection.cpp:94]: YES it’s great

**Comment by DHowett-MSFT** [comment]
Alright, got to this:
image


default domain _if available_, ID otherwise, and this supports loading up old version credentials

**Pull request merged**